reactor-test/src/main/java/reactor/test/DefaultStepVerifierBuilder.java

```diff
@@ -2411,8 +2412,11 @@ static String formatFusionMode(int m) {
 
 				Context c = Flux.fromStream(verifierSubscriber.parents())
```

Comment:
I think I have come up with a better foolproof way of asserting the right Context (ie, the Context writes, from the perspective of the leftmost operator/the source):

```java
		@Override
		public StepVerifier.Step<T> then() {
			return step.consumeSubscriptionWith(s -> {
				//the current subscription
				Scannable lowest = Scannable.from(s);

				//if it is not Scannable, then we have an alien Publisher probably
				if (!lowest.isScanAvailable()) {
					this.contextExpectations.accept(null);
					return;
				}

				//attempt to go back to the leftmost parent to check the Context from its perspective
				Context c = Flux.<Scannable>
						fromStream(lowest.parents())
						.takeLast(1)
						.singleOrEmpty()
						//no parent? must be a ScalaSubscription or similar source
						.switchIfEmpty(Mono.just(lowest))
						//unless it is directly a CoreSubscriber, let's try to scan the leftmost, see if it has an ACTUAL
						.map(sc -> (sc instanceof CoreSubscriber) ? sc : sc.scan(Scannable.Attr.ACTUAL))
						//we are ultimately only interested in CoreSubscribers' Context
						.ofType(CoreSubscriber.class)
						.map(CoreSubscriber::currentContext)
						//if it wasn't a CoreSubscriber (eg. custom or vanilla Subscriber) there won't be a Context
						.block();

				this.contextExpectations.accept(c);
			});
		}
```

This would come with the following tests in `DefaultContextExpectationsTest`:

```java

	@Test
	public void contextAccessibleEmptyInChain() {
		assertContextExpectation(s -> s, e -> e.hasSize(0));
	}

	@Test
	public void notContextAccessibleDueToPublisher() {
		Publisher<Integer> publisher = subscriber -> subscriber.onSubscribe(new Subscription() {
			@Override
			public void request(long l) {
				subscriber.onNext(100);
				subscriber.onNext(101);
				subscriber.onNext(102);
				subscriber.onNext(103);
				subscriber.onNext(104);
				subscriber.onNext(105);
				subscriber.onNext(106);
				subscriber.onNext(107);
				subscriber.onNext(108);
				subscriber.onNext(109);
				subscriber.onComplete();
			}

			@Override
			public void cancel() {
				//NO-OP
			}
		});

		assertContextExpectationFails(s -> publisher, e -> e)
				.withMessage("No propagated Context");
	}
```

Reply:
(note the second tests replaces `notContextAccessible`)